notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+       "2023-10-23T00:00:00.000Z Japan          Magic 8-ball               5   \n",
+       "                                        Pet rock                  16   \n",
+       "                                        Slinky                    11   \n",
+       "\n",
+       "                                                      forecast_quantity  \n",
+       "week_start               client_country product                          \n",
+       "2023-10-16T00:00:00.000Z Japan          Magic 8-ball                949  \n",
+       "                                        Pet rock                   1004  \n",
+       "                                        Slinky                      469  \n",
+       "2023-10-23T00:00:00.000Z Japan          Magic 8-ball                687  \n",
+       "                                        Pet rock                    745  \n",
+       "                                        Slinky                      362  "
+      ]
+     },
+     "execution_count": 31,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# we can read data independently from each source, and join it in python\n",
+    "start_at = print_timestamp()\n",
+    "sql = '''\n",
+    "    SELECT \n",
+    "       TIME_FLOOR(__time, 'P1W') week_start, \n",
+    "       \"client_country\",\n",
+    "       \"product\",\n",
+    "       SUM(\"quantity\") \"sales_quantity\"\n",
+    "    FROM \"sales\"\n",
+    "    WHERE __time > TIME_FLOOR( CURRENT_TIMESTAMP, 'P1D') - INTERVAL '7' DAY\n",
+    "    AND \"product\" in ('Slinky', 'Magic 8-ball', 'Pet rock')\n",
+    "    AND \"client_country\" = 'Japan'\n",
+    "    GROUP BY 1,2,3\n",
+    "'''\n",
+    "sales_data = pandas.json_normalize(sql_client.sql(sql)).set_index(['week_start','client_country','product'])\n",
+    "\n",
+    "sql = '''\n",
+    "    SELECT \n",
+    "       TIME_FLOOR(__time, 'P1W') week_start, \n",
+    "       \"client_country\",\n",
+    "       \"product\",\n",
+    "       SUM(\"quantity\") \"forecast_quantity\"\n",
+    "    FROM \"sales_forecast\"\n",
+    "    WHERE __time > TIME_FLOOR( CURRENT_TIMESTAMP, 'P1D') - INTERVAL '7' DAY\n",
+    "    AND \"product\" in ('Slinky', 'Magic 8-ball', 'Pet rock')\n",
+    "    AND \"client_country\" = 'Japan'\n",
+    "    GROUP BY 1,2,3\n",
+    "'''\n",
+    "sales_forecast_data = pandas.json_normalize(sql_client.sql(sql)).set_index(['week_start','client_country','product'])\n",
+    "\n",
+    "# we join it here on the \"application\" side:\n",
+    "result = sales_data.join(sales_forecast_data, on=['week_start','client_country','product'])\n",
+    "end_at=print_timestamp()\n",
+    "print(f\"Duration: {(end_at-start_at).total_seconds()*1000} ms \")\n",
+    "result"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c0441a7f-4a96-4ffe-9f19-11ecb6f06b22",
+   "metadata": {},
+   "source": [
+    "### Single Query with UNION ALL - Sales and Forecast in Different Result Rows\n",
```

Comment:
This section could be added to the UNION ALL notebook for sure.

Reply:
Yes. I think these are valuable techniques for Druid.

Reply:
But I also think it is useful content in the context of planning query patterns and modeling, so I think it is valuable in both places.